File: EPI/19.3/swift/19.3.swift
public enum Color: Int {
    case White = 0
    case Black = 1
}

private typealias Coord = (row: Int, col: Int)

// Fancy extension for using `grid[coord]` instead of `grid[coord.row][coord.col]`.
private extension Array where Element: MutableCollection {
    subscript(tuple: (Index, Element.Index)) -> Element.Iterator.Element  {
        get {
            return self[tuple.0][tuple.1]
        }
        set(newValue) {
            self[tuple.0][tuple.1] = newValue
        }
    }
}

public func fillEnclosed(_ grid: inout [[Color]]) {
    var visited = grid.map { $0.map { $0 == .Black } }

    for row in 0..<grid.count {
        for col in 0..<grid[row].count {
            let coord = (row, col)

            if !visited[coord] {
                let enclosed = visitRegion(coord, in: grid, visited: &visited)
                if enclosed {
                    fillRegion(coord, in: &grid)
                }
            }
        }
    }
}

private func visitRegion(_ coord: Coord, in grid: [[Color]], visited: inout [[Bool]]) -> Bool {
    visited[coord] = true

    // Do not return early here. All cells in the region need to be marked as visited.
    var enclosed = !onBorder(coord, in: grid)

    for neighbor in neighbors(coord, in: grid) where !visited[coord] {
        if !visitRegion(neighbor, in: grid, visited: &visited) {
            enclosed = false
        }
    }

    return enclosed
}

private func fillRegion(_ coord: Coord, in grid: inout [[Color]]) {
    grid[coord] = .Black

    for neighbor in neighbors(coord, in: grid) where grid[neighbor] == .White {
        fillRegion(neighbor, in: &grid)
    }
}

private func onBorder(_ coord: Coord, in grid: [[Color]]) -> Bool {
    return coord.row == 0 || coord.row == grid.count - 1 ||
        coord.col == 0 || coord.col == grid[coord.row].count - 1
}

private func neighbors(_ coord: Coord, in grid: [[Color]]) -> [Coord] {
    let (row, col) = coord

    let potentialNeighbors = [
        (row - 1, col),
        (row + 1, col),
        (row, col - 1),
        (row, col + 1),
    ]

    return potentialNeighbors.filter { (r, c) in
        r >= 0 && r < grid.count &&
        c >= 0 && c < grid[r].count
    }
}

File: EPI/19.3/typescript/19.3.ts
interface Cell {
    row: number;
    col: number;
}

export enum Color {
    WHITE,
    BLACK,
}

export function fillEnclosed(grid: Color[][]): void {
    const visited = grid.map(colors => colors.map(color => color === Color.BLACK));

    grid.forEach((colors, row) => colors.forEach((_, col) => {
        if (row === 0 || row === grid.length - 1 || col === 0 || col === grid[row].length - 1) {
            dfs(visited, {row, col});
        }
    }));

    visited.forEach((values, row) => values.forEach((value, col) => {
        if (!value) {
            grid[row][col] = Color.BLACK;
        }
    }));
}

function dfs(visited: boolean[][], cell: Cell): void {
    if (visited[cell.row][cell.col]) {
        return;
    }

    visited[cell.row][cell.col] = true;

    for (const neighbor of neighbors(visited, cell)) {
        dfs(visited, neighbor);
    }
}

const neighbors = (grid: any[][], {row, col}: Cell) => [
    {row: row - 1, col},
    {row, col: col + 1},
    {row: row + 1, col},
    {row, col: col - 1},
].filter(({row: r, col: c}) =>
    r >= 0 && r < grid.length &&
    c >= 0 && c < grid[r].length
);
